cluster-autoscaler/estimator/binpacking_estimator.go

```diff
@@ -0,0 +1,116 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package estimator
+
+import (
+	"sort"
+
+	"k8s.io/contrib/cluster-autoscaler/simulator"
+	kube_api "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/resource"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache"
+)
+
+// podInfo contains Pod and score that corresponds to how important it is to handle the pod first.
+type podInfo struct {
+	score float64
+	pod   *kube_api.Pod
+}
+
+type byScoreDesc []*podInfo
+
+func (a byScoreDesc) Len() int           { return len(a) }
+func (a byScoreDesc) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byScoreDesc) Less(i, j int) bool { return a[i].score > a[j].score }
+
+// BinpackingNodeEstimator estimates the number of needed nodes to handle the given amount of pods.
+type BinpackingNodeEstimator struct {
+	predicateChecker *simulator.PredicateChecker
+}
+
+// NewBasicNodeEstimator builds BasicNodeEstimator.
+func NewBinpackingNodeEstimator(predicateChecker *simulator.PredicateChecker) *BinpackingNodeEstimator {
+	return &BinpackingNodeEstimator{
+		predicateChecker: predicateChecker,
+	}
+}
+
+// Estimate implements First Fit Decreasing bin-packing approximation algorithm.
+// See https://en.wikipedia.org/wiki/Bin_packing_problem for more details.
+// While it is a multi-dimensional bin packing (cpu, mem, ports) in most cases the main dimension
+// will be cpu thus the estimated overprovisioning of 11/9 * optimal + 6/9 should be
+// still be maintained.
+// It is assumed that all pods from the given list can fit to nodeTemplate.
+func (estimator *BinpackingNodeEstimator) Estimate(pods []*kube_api.Pod, nodeTemplate *schedulercache.NodeInfo) int {
+	podInfos := make([]*podInfo, 0, len(pods))
+
+	// Calculate score for all pods.
+	// Score is defined as cpu_sum/node_capacity + mem_sum/node_capacity. With this pods that
+	// have bigger requirements are processed first.
+	for _, pod := range pods {
+		cpuSum := resource.Quantity{}
+		memorySum := resource.Quantity{}
+
+		for _, container := range pod.Spec.Containers {
+			if request, ok := container.Resources.Requests[kube_api.ResourceCPU]; ok {
+				cpuSum.Add(request)
+			}
+			if request, ok := container.Resources.Requests[kube_api.ResourceMemory]; ok {
+				memorySum.Add(request)
+			}
+		}
+		score := float64(0)
+		if cpuCapcaity, ok := nodeTemplate.Node().Status.Capacity[kube_api.ResourceCPU]; ok && cpuCapcaity.MilliValue() > 0 {
+			score += float64(cpuSum.MilliValue()) / float64(cpuCapcaity.MilliValue())
+		}
+		if memCapcaity, ok := nodeTemplate.Node().Status.Capacity[kube_api.ResourceMemory]; ok && memCapcaity.Value() > 0 {
+			score += float64(memorySum.Value()) / float64(memCapcaity.Value())
+		}
+
+		podInfos = append(podInfos, &podInfo{
+			score: score,
+			pod:   pod,
+		})
+	}
+	sort.Sort(byScoreDesc(podInfos))
+
+	nodeWithPod := func(nodeInfo *schedulercache.NodeInfo, pod *kube_api.Pod) *schedulercache.NodeInfo {
```

Comment:
Please add a comment:

"nodeWithPod function returns NodeInfo, which is a copy of nodeInfo argument with an additional pod scheduled on it"